tower-http/src/compression/body.rs

```diff
@@ -1,320 +1,233 @@
-                CompressionBody {
-                    inner: BodyInner::Brotli(framed_read),
-                }
-            }
-            Encoding::Identity => {
-                return Response::from_parts(
-                    parts,
-                    CompressionBody {
-                        inner: BodyInner::Identity(body, PhantomData),
-                    },
-                )
-            }
-        };
-
-        parts.headers.remove(header::CONTENT_LENGTH);
-
-        parts.headers.insert(
-            header::CONTENT_ENCODING,
-            HeaderValue::from_static(encoding.to_str()),
-        );
-
-        http::Response::from_parts(parts, body)
-    }
+#[pin_project(project = BodyInnerProj)]
+pub(crate) enum BodyInner<B>
+where
+    B: Body,
+{
+    #[cfg(feature = "compression-gzip")]
+    Gzip(#[pin] WrapBody<GzipEncoder<IoBody<B>>>),
+    #[cfg(feature = "compression-deflate")]
+    Deflate(#[pin] WrapBody<ZlibEncoder<IoBody<B>>>),
+    #[cfg(feature = "compression-br")]
+    Brotli(#[pin] WrapBody<BrotliEncoder<IoBody<B>>>),
+    Identity(#[pin] IoBody<B>),
 }
 
 impl<B> Body for CompressionBody<B>
 where
     B: Body,
 {
     type Data = Bytes;
-    type Error = Error<B::Error>;
+    type Error = io::Error;
 
-    #[allow(unused_variables, unreachable_code)]
     fn poll_data(
         self: Pin<&mut Self>,
         cx: &mut Context<'_>,
     ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
-        let (bytes, inner): (_, Pin<&mut Adapter<B, <B as Body>::Error>>) =
-            match self.project().inner.project() {
-                #[cfg(feature = "compression-gzip")]
-                BodyInnerProj::Gzip(mut framed_read) => (
-                    ready!(framed_read.as_mut().poll_next(cx)),
-                    framed_read.get_pin_mut().get_pin_mut().get_pin_mut(),
-                ),
-                #[cfg(feature = "compression-deflate")]
-                BodyInnerProj::Deflate(mut framed_read) => (
-                    ready!(framed_read.as_mut().poll_next(cx)),
-                    framed_read.get_pin_mut().get_pin_mut().get_pin_mut(),
-                ),
-                #[cfg(feature = "compression-br")]
-                BodyInnerProj::Brotli(mut framed_read) => (
-                    ready!(framed_read.as_mut().poll_next(cx)),
-                    framed_read.get_pin_mut().get_pin_mut().get_pin_mut(),
-                ),
-                BodyInnerProj::Identity(inner, _) => {
-                    return match ready!(inner.poll_data(cx)) {
-                        Some(Ok(mut data)) => {
-                            Poll::Ready(Some(Ok(data.copy_to_bytes(data.remaining()))))
-                        }
-                        Some(Err(e)) => Poll::Ready(Some(Err(Error::Body(e)))),
-                        None => Poll::Ready(None),
-                    };
+        match self.project().0.project() {
+            #[cfg(feature = "compression-gzip")]
```

Comment:
Do we have a way in CI right now to check all these variations of feature flags?

Reply:
Yes https://github.com/tower-rs/tower-http/blob/master/.github/workflows/CI.yml#L46. Learned about this from tower's GH actions setup.